kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h

```diff
@@ -0,0 +1,614 @@
+        ProcessInfo &rCurrentProcessInfo)
+    {
+        //filling with zero the matrix (creating the structure)
+        Timer::Start("MatrixStructure");
+
+        const std::size_t equation_size = BaseType::mDofSet.size();
+
+#ifdef USE_GOOGLE_HASH
+        std::vector<google::dense_hash_set<std::size_t>> indices(equation_size);
+        const std::size_t empty_key = 2 * equation_size + 10;
+#else
+        std::vector<std::unordered_set<std::size_t>> indices(equation_size);
+#endif
+
+#pragma omp parallel for firstprivate(equation_size)
+        for (int iii = 0; iii < static_cast<int>(equation_size); iii++)
+        {
+#ifdef USE_GOOGLE_HASH
+            indices[iii].set_empty_key(empty_key);
+#else
+            indices[iii].reserve(40);
+#endif
+        }
+        Element::EquationIdVectorType ids(3, 0);
+
+        const int nelements = static_cast<int>(rElements.size());
+#pragma omp parallel for firstprivate(nelements, ids)
+        for (int iii = 0; iii < nelements; iii++)
+        {
+            typename ElementsContainerType::iterator i_element = rElements.begin() + iii;
+            (i_element)->EquationIdVector(ids, rCurrentProcessInfo);
+
+            // Modifying the equation IDs of this element to suit MPCs
+            this->ModifyEquationIdsForConstraints(*(*(i_element.base())), ids, rCurrentProcessInfo, mpConstraintVector);
+
+            for (std::size_t i = 0; i < ids.size(); i++)
+            {
+#ifdef _OPENMP
+                omp_set_lock(&(BaseType::mlock_array[ids[i]]));
+#endif
+                auto &row_indices = indices[ids[i]];
+                row_indices.insert(ids.begin(), ids.end());
+
+#ifdef _OPENMP
+                omp_unset_lock(&(BaseType::mlock_array[ids[i]]));
+#endif
+            }
+        }
+        const int nconditions = static_cast<int>(rConditions.size());
+#pragma omp parallel for firstprivate(nconditions, ids)
+        for (int iii = 0; iii < nconditions; iii++)
+        {
+            typename ConditionsArrayType::iterator i_condition = rConditions.begin() + iii;
+            (i_condition)->EquationIdVector(ids, rCurrentProcessInfo);
+            // Modifying the equation IDs of this element to suit MPCs
+            this->ModifyEquationIdsForConstraints(*(*(i_condition.base())), ids, rCurrentProcessInfo, mpConstraintVector);
+
+            for (std::size_t i = 0; i < ids.size(); i++)
+            {
+#ifdef _OPENMP
+                omp_set_lock(&(BaseType::mlock_array[ids[i]]));
+#endif
+                auto &row_indices = indices[ids[i]];
+                row_indices.insert(ids.begin(), ids.end());
+#ifdef _OPENMP
+                omp_unset_lock(&(BaseType::mlock_array[ids[i]]));
+#endif
+            }
+        }
+        //count the row sizes
+        unsigned int nnz = 0;
+        for (unsigned int i = 0; i < indices.size(); i++)
+            nnz += indices[i].size();
+
+        A = boost::numeric::ublas::compressed_matrix<double>(indices.size(), indices.size(), nnz);
```

Comment:
I think here you could use [CompressedMatrix](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/includes/ublas_interface.h#L90)
This way it will be easier later to remove boost

Reply:
Not my implementation here . This is also like this in residual_based_block_builder_and_solver.h

Reply:
I will change when this is changed in the base builder and solver.